### ISSUE TITLE
大佬，这个期待好久了！！！

### ISSUE THREAD

**Issue by ozy20** [JessYanCoding/ArmsComponent#1]

**Comment by JessYanCoding** [comment]
对, 其实我两个月前就开始准备了,  为了让文档和代码更好的呈现给大家, 让大家更好理解, 不断优化了很多次, 所以现在才放出来

**Comment by cocowobo** [comment]
来吧，群主生拉硬拽的把我拖上了前进的列车，并且群主还是老司机，车门已经焊死，只能继续学习前进了

**Comment by guixin567** [comment]
向大佬致敬

**Comment by YXY1109** [comment]
公司业务越来越多，之前的管理方式已经不行了。看来组件化要多多了解了，向大佬致敬

**Comment by Samuel12316** [comment]
來朝聖大佬，從入門到現在的組件化都是看大佬的demo跟文章。

**Comment by JessYanCoding** [comment]
@Samuel12316 ArmsComponent 再添神器, [一键搭建 ArmsComponent 快速组件化方案的整体组件架构, 让新手也可以一秒开始组件化项目, 开发神器, 让您一秒起飞, 避免组件化从入门到放弃!](https://github.com/JessYanCoding/ArmsComponent-Template)

**Comment by zgxzgxzg** [comment]
开源，分享，促进行业发展~

**Comment by flybirdxx** [comment]
如果我自己需要一个Application，正确方式应该是在哪新建初始化？

**Comment by LatoAndroidw** [comment]
请问下，关于跨组件通信，作者有没有计划在arouter外做关于多个不同组件apk的通信，例如Aidl，Socket等

**Comment by JessYanCoding** [comment]
@LatoAndroidw 暂时没有考虑，关于通信我都把它归为属于路由框架的功能，在文档中我也明确指出，如果 ARouter 不能满足需求，可以自行切换为其他可以满足需求的路由框架，框架并没有规定必须使用 ARouter，专业是的事交给专业的人去做，这方面我并不专业，而且也有很多优秀的框架存在，所以关于模块间的路由和通信我更倾向于用更专业的三方库，ArmsComponent  的优势其实是在于基于 Arms 基础库而搭建起的规范，所以我出了一键生成组件的模板，后续也会围绕着这方面来迭代，市面上大部分的路由方案都的核心点都是基于路由框架而搭建，比如 **CC**，所以你有这方面疑问我也理解

**Comment by m122469119** [comment]
组件与组件debug期间,任意组合,案例中并没有体现.

**Comment by JessYanCoding** [comment]
@m122469119 debug 期间，每个组件都是独立的 app，打包成单独的 apk, 你想 apk 之间怎么组合，这个不是插件化方案，只可能在宿主中根据自己的需求在多个组件中挑选出自己需要的某几个组件进行依赖，然后一起打包成一个 apk

**Comment by JessYanCoding** [comment]
你说的意思，其实将直播组件作为宿主，不需要单独建宿主，你自己改 build.gradle 让这个组件成为宿主依赖其他组件就可以了

爱奇艺开源的跨进程组件通信框架 [Andromeda](https://github.com/iqiyi/Andromeda/blob/master/CHINESE_README.md)

**Comment by hdsb2006** [comment]
方法数超的解决方案有好的吗，如果在app中加的话，其他moduel单独运行的时候也会报方法数超的问题。如果在moduel中加入 multiDexEnabled = true这类的解决方案的话，会报异常。

**Comment by hdsb2006** [comment]
请问组件化如何添加混淆？添加混淆后，注入好像失效了

**Comment by BobEve** [comment]
你好，之前接入时遇到一个问题：在自定义View中触发了System.err的错误，但是框架捕获异常后按照Verbose打印到控制台，检查了好几次才发现。建议将System.err的异常打印为设置为error
![image](https://user-images.githubusercontent.com/5792599/45690125-37bed600-bb88-11e8-9e5d-d037297b5c7b.png)

**Comment by JessYanCoding** [comment]
这个日志不是框架打印的

**Comment by LanJD** [comment]
您好，有没有具体一点的组件化项目demo可以学习，最好都涉及了文章里提到的一些模块，不然看完下来还是概念还是很模糊 捂脸。。。 @JessYanCoding

**Comment by zhangcongmin** [comment]
![image](https://user-images.githubusercontent.com/8501196/47981049-47c55180-e105-11e8-8737-732b6a3465a8.png)
不同版本多出好多相同的库？这个怎么解决

**Comment by JessYanCoding** [comment]
@zhangcongmin 目前 MVPArms v2.4.1 使用的是 Support 包是 27.1.1 但 ArmsComponent 却将 Support 包更新为了 v28.0.0，这段时间 MVPArms 更新了新版本，就会统一使用 v28.0.0

**Comment by tydqcjj** [comment]
大神你好！我对组件化的理解是我们玩的棋牌游戏（或者宝宝巴士）。进app就是登陆和各种游戏菜单（比如有斗地主、麻将、斗牛等），这个主界面就是壳工程，点斗地主进入的是一个module，这个module是可以单独编译成一个斗地主apk的，也可以作为一个library和其他麻将等编译成一个聚合apk。
可是我想把我游戏盒子项目组件化，可是我的每一个模块是一个fragment，也就是类似微信的结构，通过底部导航来切换四个fragment，请问怎么组件化，fragment必须依赖activity啊，怎么分离？

**Comment by JessYanCoding** [comment]
由于 ArmsComponent 在 组件单独运行时 和 组件集成宿主时使用了两个不同的 Manifest，所以在组件单独运行时，必须要有一个专门用于测试的宿主 Activity，用于你这个组件的 Fragment 依赖，这个宿主 
 Activity 只在 组件单独运行时 的 Manifest 中声明，在组件集成宿主时 的 Manifest 中不需要声明也不需要打包进去，只用于测试，这是你这个组件就可以单独运行了

当集成宿主时，通过 ARouter 将这个组件的 Fragment 暴露给其他组件或宿主并依附到需要的 Activity 下即可，思路是这样，具体代码自己思考把

**Comment by z2700589** [comment]
大神你好,如果用到了第三方sdk的so文件,这些文件放在那个module下合适呢? 调试了好久发现只能放在宿主app下,放到res或者sdk下都会报More than one file was found with OS independent path 'lib/armeabi-v7a/libruntimecore_java.so'这个错误.求大神指点

**Comment by LanJD** [comment]
大佬你好，什么时候出个kotlin代码的框架呢？

**Comment by JessYanCoding** [comment]
@LanJD 目前没这个想法

**Comment by hqye** [comment]
@JessYanCoding 亲，targetSdkVersion为28的话要注意network_security_config这个问题啊，我用模拟器不加这个访问不了接口

**Comment by JessYanCoding** [comment]
@hqye 感谢提醒，最近比较忙，能否把你已经配置完成的 network_security_config 贴出来，如果能将代码 [pr](https://github.com/JessYanCoding/ArmsComponent/compare) 到仓库那就最好不过了

**Comment by hqye** [comment]
@JessYanCoding  我才看到，(●'◡'●)你都已经改好了

**Comment by z2700589** [comment]
Type me.jessyan.armscomponent.commonsdk.base.IBaseActivity_MembersInjector is defined multiple times: F:\AndroidProjects\AppRepository\MobileSystem\JNRespiratory\module_system\build\intermediates\runtime_library_classes\release\classes.jar:me/jessyan/armscomponent/commonsdk/base/IBaseActivity_MembersInjector.class, F:\AndroidProjects\AppRepository\MobileSystem\JNRespiratory\module_replace\build\intermediates\runtime_library_classes\release\classes.jar:me/jessyan/armscomponent/commonsdk/base/IBaseActivity_MembersInjector.class


因为业务需要  重写了baseActivity，但是打包的时候说这个类被定义了多次，改怎么解决呢？

**Comment by JessYanCoding** [comment]
@z2700589 https://github.com/JessYanCoding/MVPArms/pull/341

**Comment by z2700589** [comment]
大佬，组件化的混淆应该怎么配置，是只需要在宿主层配置proguard-rules.pro，还是每个module都需要配置？

**Comment by JessYanCoding** [comment]
@z2700589 宿主，你就把 Module 当成三方库
